Improves CI test resource cleanup to prevent orphaned AWS resources from failed/cancelled test runs.

sean-navapbc opened the pull request:
Summary

  Improves CI test resource cleanup to prevent orphaned AWS resources from failed/cancelled test runs.

  Changes

  1. Cleanup Script (template-only-bin/cleanup-test-resources)

  New comprehensive script to find and delete orphaned test resources tagged with plt-tst-act-*:
  - Route 53 hosted zones
  - ECS clusters, services, and task definitions
  - Load balancers and target groups
  - S3 buckets and DynamoDB tables
  - KMS keys (scheduled for deletion with 7-day wait)
  - Supports --dry-run mode to preview changes

  2. Destroy Script (template-only-bin/destroy-app-service)

  Enhanced teardown script with ECS task definition cleanup:
  - Task definitions aren't auto-deleted by Terraform (https://github.com/hashicorp/terraform-provider-aws/issues/5919)
  - Uses Resource Groups Tagging API to find task definitions by project tag
  - Gets project config from terraform outputs instead of inferring from pwd

  3. Scan Workflow (.github/workflows/template-only-scan-orphaned-infra-test-resources.yml)

  Daily scheduled workflow to detect orphaned resources:
  - Runs cleanup script in dry-run mode
  - Fails if orphaned resources are found (alerts on resource leaks)

  4. Cleanup Workflow (.github/workflows/template-only-cleanup-orphaned-infra-test-resources.yml)

  Manual workflow to clean up orphaned resources when needed.

  Test Plan

  - Run cleanup script with --dry-run to verify detection
  - Run cleanup script to clear existing orphaned resources
  - Verify CI tests pass
  - Monitor for orphaned resources after merge

doshitan commented on template-only-test/template_infra_test.go:
```diff
@@ -191,40 +191,52 @@ func ValidateDevEnvironment(t *testing.T) {
 func TeardownAccount(t *testing.T) {
 	fmt.Println("::group::Destroying account resources")
-	shell.RunCommand(t, shell.Command{
+	err := shell.RunCommandE(t, shell.Command{
```
As the resources generally need to be torn down in a particular order, I'm not sure trying to carry on makes sense.

I could see value in potentially retrying each destroy step though. If a step hits some AWS timeout. Not sure how common that would be though. @sean-navapbc does it seem worth the effort to add retry logic?

sean-navapbc replied:
Yeah it does seem worth it. I'll add some

doshitan commented on bin/cleanup-test-resources:
```diff
@@ -0,0 +1,223 @@
+#                             If not provided, finds all matching projects
+#
+# Options:
+#   --dry-run        - List resources that would be deleted without deleting them
```
Doesn't have to be now, but could be useful to have a periodic GH Action that basically runs this and notifies if there's any resources, kinda like the `scan-orphaned-environments.yml` and `check-infra-deploy-status.yml` workflows do for project instances.

sean-navapbc commented on .github/workflows/template-only-ci-infra.yml:
```diff
@@ -109,3 +109,20 @@ jobs:
           IMAGE_TAG: ${{ github.sha }}
           PROJECT_NAME: ${{ env.PROJECT_NAME }}
+
+      # Cleanup step that runs even if tests fail, timeout, or are cancelled
```
I've created the cleanup-ophaned-test-resources github action that can be dispatched.

doshitan commented on .github/workflows/cleanup-orphaned-test-resources.yml:
Name should start with `template-only-`

doshitan commented on .github/workflows/scan-orphaned-test-resources.yml:
```diff
@@ -0,0 +1,56 @@
+name: Scan Orphaned Test Resources
```
```suggestion
name: Template Scan Orphaned Infra Test Resources
```

sean-navapbc commented:
Re: the question about go test timeouts - I reviewed the codebase and workflow runs but couldn't find concrete evidence that the 30m go test timeout has been hit in practice. 

The PR description mentions `plt-tst-act-32078` from Oct 20 as an example of orphaned resources, but it's unclear if that was specifically due to a go test timeout or another cause (cancellation, failure, etc.).

I've already addressed the other review comments:
-  Fixed teardown error handling (changed from DoWithRetryE to DoWithRetry)
- Created runCommandWithRetry utility function
- Removed GitHub Actions cleanup step

For the timeout question: if we haven't actually seen the go test timeout being hit, the current 30m timeout + the resilient teardown logic (with retry) should be sufficient. The GH Actions job timeout would still serve as the backup catch-all for truly hanging processes.

Happy to make additional changes if there's evidence this has been a problem in practice.

doshitan commented on template-only-bin/template-only-cleanup-test-resources:
```diff
@@ -0,0 +1,225 @@
+
+  if ! aws s3api head-bucket --bucket "${bucket_name}" 2>/dev/null; then
+    echo "No terraform state bucket found for ${project} (likely already cleaned up)"
+    return 0
```
If there are resources that exist tagged with a project, but the project no longer has a state bucket, those resources would still need cleaned up. Since we are not using terraform to do the cleanup, this wouldn't be a blocker. Is the problem that the resource tagging API sometimes returns phantom resources?

sean-navapbc replied:
Good catch.That is correct, if the state bucket was deleted but resources remain, we should still attempt cleanup since we're using AWS CLI directly rather than terraform.

Updated:  the state bucket check is now informational only. The script will continue to search for and clean
   up resources even if the bucket is missing.

sean-navapbc commented:
> Interesting, looks like AWS didn't support deleting ECS Task Definitions [until (relatively) recently](https://aws.amazon.com/blogs/containers/announcing-amazon-ecs-task-definition-deletion/).
> 
> Until [support lands in terraform itself](https://github.com/hashicorp/terraform-provider-aws/issues/29682), we could add some logic in `/template-only-bin/destroy-app-service` to clean up the task definition created for the run (terraform should deregister the task definition), as this is probably the only place that cares at the moment. We could also have a script (or a part of the general cleanup script) to just delete all inactive task definitions.
> 
> Doesn't have to be done here, but if not, we should create a ticket for the followup.

Let's fix it here. Why not

doshitan commented on .github/workflows/template-only-scan-orphaned-infra-test-resources.yml:
```diff
@@ -0,0 +1,60 @@
+          # Suppress broken pipe errors from early pipe termination
+          if (echo "$output" | grep -q "Found [1-9][0-9]* resources") 2>/dev/null; then
+            # Extract resource count and project names for notification
+            resource_info=$(echo "$output" | grep -E "(Found [0-9]+ resources|Cleaning up project:|plt-tst-act-)" | head -20) 2>/dev/null
```
This looks to be erroring?

```
=== Cleanup complete ===
grep: write error: Broken pipe
/home/runner/work/_temp/f7e47898-e742-4bd9-ae9b-997ba22d2e5e.sh: line 9: echo: write error: Broken pipe
```

https://github.com/navapbc/template-infra/actions/runs/19839044285/job/56843173457?pr=973

doshitan commented on .github/workflows/template-only-scan-orphaned-infra-test-resources.yml:
```diff
@@ -0,0 +1,60 @@
+    name: " "
+    needs: scan
+    if: failure()
+    uses: ./.github/workflows/send-system-notification.yml
```
This only works for template instances.

We could make a `.github/workflows/template-only-*` version that's hardcoded, but can punt on this for now I think. Easy enough to just manually check in on the repo every so often.

doshitan reviewed (commented):
Seems like we are getting close. I think we can run `.github/workflows/template-only-cleanup-orphaned-infra-test-resources.yml` for a specific project or two (maybe like `plt-tst-act-n48mbv` and `plt-tst-act-01485` covering medium and small cases) and post evidence of it working, with before/after snips of the relevant "scan" output.

doshitan reviewed (commented):
Lint issue: https://github.com/navapbc/template-infra/actions/runs/20067415640/job/57559649603?pr=973

The [latest scan run](https://github.com/navapbc/template-infra/actions/runs/20067414624/job/57559646079) lists a bunch of projects found, but no resources for them and some broken pipe errors? I still see a bunch of inactive task definitions at least. So something broke.

@sean-navapbc you mentioned doing some testing of the cleanup script, can you post that here? What was tested and what was the outcome.

sean-navapbc commented on template-only-bin/cleanup-test-resources:
```diff
@@ -0,0 +1,248 @@
+      for task_def_arn in ${task_definitions}; do
+        aws ecs deregister-task-definition --region "${AWS_REGION}" --task-definition "${task_def_arn}" || echo "Failed to deregister ${task_def_arn}"
+      done
+    fi
```
So both scripts now have the complete task definition cleanup:
  1. cleanup-test-resources (lines 208-212) - for cleaning up orphaned resources
  2. destroy-app-service (lines 75-80) - for test runs to clean up after themselves

  Both scripts properly delete ACTIVE and INACTIVE task definitions.

sean-navapbc commented:
---
  Issues Fixed

  1. Lint issue (SC2001) ✅

  Fixed in commit: a5d77ae

  Changed from:
  bucket_name=$(echo "${s3_arn}" | sed 's/arn:aws:s3::://')
  To:
  bucket_name="${s3_arn#arn:aws:s3:::}"

  Evidence: https://github.com/navapbc/template-infra/actions/runs/20143437468/job/57817201077

  2. Broken pipe errors ✅

  Fixed in commit: aea90f5

  Added stderr redirect to suppress broken pipe messages when head terminates early:
  { echo "$output" | head -50 || true; } 2>/dev/null

  Evidence: https://github.com/navapbc/template-infra/actions/runs/20143436251/job/57817196876

  3. Inactive task definitions not being found ✅

  Fixed in commit: daa0f2e

  Added cleanup_inactive_task_definitions() function that directly queries ECS for inactive task definitions matching plt-tst-act-*
   pattern, bypassing the Resource Groups Tagging API which doesn't return INACTIVE task definitions.

  aws ecs list-task-definition-families --family-prefix "plt-tst-act-" --status INACTIVE
  aws ecs delete-task-definitions --task-definitions "${task_arn}"

  CI Status

  All checks passing:
  - Lint template scripts ✅
  - Scan for orphaned test resources ✅ (no broken pipe errors)
  - tfsec, checkov, terraform format ✅

doshitan commented on infra/test/helpers.go:
```diff
@@ -18,3 +21,10 @@ import (
 }
+
+// runCommandWithRetry runs a shell command with retry logic
+func runCommandWithRetry(t *testing.T, description string, maxRetries int, sleepBetweenRetries time.Duration, command shell.Command) {
```
This isn't used in these tests, should probably drop for now?

doshitan commented on template-only-bin/destroy-app-service:
```diff
@@ -27,3 +27,60 @@ terraform init -reconfigure -backend-config="${backend_config_file}"
+cd ../../..
+
+# Get the project name from the current directory or AWS resources
+project_name=$(basename "$(pwd)")
```
```suggestion
project_name=$(terraform -chdir="infra/project-config" output --raw project_name)
```

doshitan commented:
The PR description doesn't quite match the current scope of the work, so can update that. A few other tweaks and probably some clearer comments/docs, but then probably ready to ship. Once we are confident things are mostly working, we should run the cleanup script, confirming everything is cleared out. Then merge and monitor how things work out going forward.

sean-navapbc commented on template-only-bin/cleanup-test-resources:
```diff
@@ -0,0 +1,372 @@
+done
+
+# Clean up orphaned inactive task definitions
+# These aren't returned by the Resource Groups Tagging API, so we query ECS directly
```
Updated:

The Resource Groups Tagging API may not return INACTIVE task definitions,
and some task definitions may remain if a previous cleanup was interrupted.
This function scans ECS directly to find any remaining task definitions
tagged with plt-tst-act-* projects.

sean-navapbc commented on template-only-bin/destroy-app-service:
```diff
@@ -27,3 +27,60 @@ terraform init -reconfigure -backend-config="${backend_config_file}"
+
+# -----------------------------------------------------------------------------
+# Clean up ECS task definitions that were created during the test run
+# Task definitions are not managed by Terraform, so they need manual cleanup
```
Modified:

Task definitions are managed by Terraform, but the AWS provider doesn't yet
support automatic deletion of task definitions on destroy.
See: https://github.com/hashicorp/terraform-provider-aws/issues/5919

sean-navapbc commented on template-only-bin/destroy-app-service:
```diff
@@ -27,3 +27,60 @@ terraform init -reconfigure -backend-config="${backend_config_file}"
+# Clean up ECS task definitions that were created during the test run
+# Task definitions are not managed by Terraform, so they need manual cleanup
+# -----------------------------------------------------------------------------
+cd ../../..
```
Done. Changed to pushd/popd pattern - cleaner and doesn't require calculating the relative path back.

sean-navapbc commented on template-only-bin/destroy-app-service:
```diff
@@ -20,10 +20,69 @@ sed -i.bak 's/force_destroy = var.is_temporary/force_destroy = true/g' infra/mod
+# Clean up ECS task definitions that were created during the test run
+# Task definitions are managed by Terraform, but the AWS provider doesn't yet
+# support automatic deletion of task definitions on destroy.
+# See: https://github.com/hashicorp/terraform-provider-aws/issues/5919
```
corrected
